File: java/src/main/java/com/example/doubleup/dao/AccessLogRepository.java
package com.example.doubleup.dao;

import com.example.doubleup.model.AccessLog;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.time.OffsetDateTime;
import java.util.UUID;

@Repository
public interface AccessLogRepository extends JpaRepository<AccessLog, UUID> {

    @Query("SELECT COUNT(e) FROM AccessLog e WHERE e.createdAt >= :startTime and ipAddress = :ipAddress")
    int countRowsForLastMinute(String ipAddress, OffsetDateTime startTime);
}

File: go/dao/access_log_repository.go
package dao

import (
    "fmt"
    "time"
    "database/sql"
    "double_up/model"
)

func CreateAccessLog(db *sql.DB, accessLog *model.AccessLog) error {
    query := "INSERT INTO access_log (id, created_at, ip_address, api) VALUES ($1, $2, $3, $4)"
    _, err := db.Exec(query, accessLog.ID, accessLog.CreatedAt, accessLog.IpAddress, accessLog.Api)
    if err != nil {
        return fmt.Errorf("AccessLog cannot be created, id: %s. Error: %s", accessLog.ID, err)
    }
    return nil
}

func CountRowsForLastMinute(db *sql.DB, ipAddress string, startTime time.Time) (int16, error) {
    query := "SELECT COUNT(id) FROM access_log WHERE created_at >= $1 and ip_address = $2"
    row := db.QueryRow(query, startTime, ipAddress)

    var result int16
	err := row.Scan(&result)
	if err != nil {
	    return 0, fmt.Errorf("Error in CountRowsForLastMinute func: %s", err)
	}
    return result, nil
}
